external/fv3fit/fv3fit/data/synthetic.py

```diff
@@ -0,0 +1,238 @@
+
+    def open_tfdataset(
+        self, local_download_path: Optional[str], variable_names: Sequence[str],
+    ) -> tf.data.Dataset:
+        """
+        Args:
+            local_download_path: if provided, cache data locally at this path
+            variable_names: names of variables to include when loading data
+        Returns:
+            dataset containing requested variables, each record is a mapping from
+                variable name to variable value, and each value is a tensor whose
+                first dimension is the batch dimension
+        """
+        dataset = get_noise_tfdataset(
+            variable_names,
+            scalar_names=self.scalar_names,
+            nsamples=self.nsamples,
+            nbatch=self.nbatch,
+            ntime=self.ntime,
+            nx=self.nx,
+            ny=self.nx,
+            nz=self.nz,
+            noise_amplitude=self.noise_amplitude,
+        )
+        if local_download_path is not None:
+            dataset = dataset.cache(local_download_path)
+        return dataset
+
+    @classmethod
+    def from_dict(cls, d: dict) -> "TFDatasetLoader":
+        return dacite.from_dict(
+            data_class=cls, data=d, config=dacite.Config(strict=True)
+        )
+
+
+@register_tfdataset_loader
+@dataclasses.dataclass
+class SyntheticWaves(TFDatasetLoader):
+    """
+    Attributes:
+        nsamples: number of samples to generate per batch
+        nbatch: number of batches to generate
+        nx: length of x- and y-dimensions to generate
+        nz: length of z-dimension to generate
+        scalar_names: names to generate as scalars instead of
+            vertically-resolved variables
+        scale_min: minimum amplitude of waves
+        scale_max: maximum amplitude of waves
+        period_min: minimum period of waves
+        period_max: maximum period of waves
+        phase_range: fraction of 2*pi to use for possible range of
+            random phase, should be a value between 0 and 1.
+        wave_type: one of "sinusoidal" or "square"
+    """
+
+    nsamples: int
+    nbatch: int
+    ntime: int
+    nx: int
+    nz: int
+    wave_type: str
+    scalar_names: List[str] = dataclasses.field(default_factory=list)
+    scale_min: float = 0.0
+    scale_max: float = 1.0
+    period_min: float = 8.0
+    period_max: float = 16.0
+    phase_range: float = 1.0
+
+    def open_tfdataset(
+        self, local_download_path: Optional[str], variable_names: Sequence[str],
+    ) -> tf.data.Dataset:
+        """
+        Args:
+            local_download_path: if provided, cache data locally at this path
+            variable_names: names of variables to include when loading data
+        Returns:
+            dataset containing requested variables, each record is a mapping from
+                variable name to variable value, and each value is a tensor whose
+                first dimension is the batch dimension
+        """
+        if self.wave_type == "sinusoidal":
```

Comment:
Nit: push this argument handling into the  `get_waves_tfdataset` function since it is already does post-processing steps with all the arguments it receives. 

Further nit: Instead of adding two new classes of synthetic data loaders, make a single class (e.g., `SyntheticData`) that has required `synthtic_data_kind` argument of type `Literal['waves','noise']`, and whatever other arguments are common to these approaches, and then also takes a bunch of kwargs appropriate to the choice of `synthtic_data_kind`

Reply:
I don't know that I can actually do this - if I push the handling into the function, it's basically going to be an if/elif statement with the existing functions inside the blocks. It's also a bit uncomfortable merging the configuration classes because there are configuration options that aren't shared between the noise types, which would get ignored depending on the settings. We can voice chat about it.